### PR TITLE
SPDataProvider addition (for S&P Market Data)

### PR DESCRIPTION
**Pull request by myrmarachne**

### CODE REVIEW

File: qf_lib/data_providers/sp_global/sp_data_provider.py [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+                                                                   func.coalesce(self.ciqiadividendchain.enddate,
+                                                                                 func.current_date())))) \
+                .join(divexrate, and_(divexrate.currencyid == self.ciqiadividendchain.currencyid,
+                                      divexrate.pricedate.between(self.ciqiadividendchain.startdate,
+                                                                  func.coalesce(self.ciqiadividendchain.enddate,
+                                                                                func.current_date())),
+                                      divexrate.pricedate == self.ciqpriceequity.pricingdate)) \
+                .filter(self.ciqtradingitem.tradingitemid.in_(tickers)) \
+                .filter(peexrate.snapid == 6) \
+                .filter(divexrate.snapid == 6) \
```
Comment (eirik-thorp) [inline qf_lib/data_providers/sp_global/sp_data_provider.py:284]: self.snapshot_id

File: qf_lib/data_providers/sp_global/sp_data_provider.py [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+        # Fetch all remaining fields            
+        field_to_fetching_method = {
+            SPField.DivYield: self._fetch_dividend_yield(tid_to_tickers.keys(), start_datetime, end_datetime),
+        }
+
+        for field in [f for f in fields if f not in price_fields]:
+            try:
+                dfs.append(field_to_fetching_method[field])
+            except KeyError:
+                raise ValueError(f"Field {field} is not supported by the SPDataProvider.") from None
```
Comment (marekbais) [inline qf_lib/data_providers/sp_global/sp_data_provider.py:139]: I'd raise immediately at the start if any of the fields are not in supported_fields, similar to frequency check

File: qf_lib/data_providers/sp_global/sp_data_provider.py [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+        into SPField objects and returns a list of all dataframes. It is assumed that the decorated function
+        returns a dataframe.
+        """
+        @wraps(fetch_func)
+        def wrapper(self, tickers, *args, **kwargs):
+            # Expecting identifiers to be the first argument
+            results = []
+            for ticker_group in grouper(1000, tickers):
+                data_frame = fetch_func(self, ticker_group, *args, **kwargs)
+                data_frame = data_frame.rename(columns={f.value: f for f in SPField})
```
Comment (eirik-thorp) [inline qf_lib/data_providers/sp_global/sp_data_provider.py:194]: Here you will fetch all values in SPField, but the dataframe will only have the fields specified in the `fields` parameter of `get_history()`. 
+ We will probably add more to the enums and the order might change etc.

File: qf_lib/data_providers/sp_global/sp_data_provider.py [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+                sql_statement = session.query(self.ciqtradingitem.tradingitemid, self.ciqcurrency.isocode) \
+                    .join(self.ciqcurrency, self.ciqtradingitem.currencyid == self.ciqcurrency.currencyid) \
+                    .filter(self.ciqtradingitem.tradingitemid.in_(ticker_group)).statement
+
+                index_col = self.ciqtradingitem.tradingitemid.key
+                currency = QFSeries(pd.read_sql(sql_statement, con=session.bind, index_col=index_col).iloc[:, 0])
+                currency = currency.rename(index=sp_id_to_ticker)
+                currencies = pd.concat([currencies, currency])
+
+        currencies = currencies.reindex(tickers).replace({np.nan: None})
```
Comment (eirik-thorp) [inline qf_lib/data_providers/sp_global/sp_data_provider.py:183]: In the `estimates` and `financials` we are changing from `None` to `np.nan`, why is it the other way round here?

File: qf_lib/data_providers/sp_global/sp_data_provider.py [hunk 5]
```diff
@@ -0,0 +1,359 @@
...
+        return currencies[0] if got_single_ticker else currencies
+
+    def supported_ticker_types(self) -> Set[Type[Ticker]]:
```
Comment (eirik-thorp) [inline qf_lib/data_providers/sp_global/sp_data_provider.py:186]: Let's make it a property

File: qf_lib/data_providers/sp_global/sp_data_provider.py [hunk 6]
```diff
@@ -0,0 +1,359 @@
...
+
+        tids_and_dates = div_adj_factor.index.intersection(tids_and_dates)
+        vals = div_adj_factor.groupby(level=0).shift(1).loc[tids_and_dates] / div_adj_factor.loc[tids_and_dates]
+        ratio = vals.groupby(level=0).cumprod().reindex_like(div_adj_factor, method="bfill") \
+            .groupby(level=0).shift(-1).fillna(1.0)
+
+        fields_to_adjust_str = {sp_field.value for sp_field in self._fields_to_adjust()}
+        fields = list(fields_to_adjust_str.intersection(data_frame.columns))
+        data_frame.loc[:, fields] = data_frame.loc[:, fields].mul(ratio['divadjfactor'], axis=0)
+        return data_frame.dropna(how="all")
```
Comment (eirik-thorp) [inline qf_lib/data_providers/sp_global/sp_data_provider.py:333]: This is a mess, let's try to rewrite a bit nicer.
Or maybe we should create a separate class for an SP data provider with the div adjusted prices?
This way we can have a minimal class to just get the prices, and then add more functionality on top of it in another class.

File: qf_lib/data_providers/sp_global/sp_data_provider.py [hunk 7]
```diff
@@ -0,0 +1,359 @@
...
+        --------
+        str, QFSeries
+            currency for the given tickers
+        """
+
+        tickers, got_single_ticker = convert_to_list(tickers, SPTicker)
+        currencies = QFSeries()
+        sp_id_to_ticker = {t.tradingitem_id: t for t in tickers}
+
+        with self._db_connection_provider.Session.begin() as session:
```
Comment (eirik-thorp) [inline qf_lib/data_providers/sp_global/sp_data_provider.py:172]: In some functions we have `session` as a input parameter to the function, and here we call `self._db_connection_provider.Session` - I think we should pick one approach for clarity.

File: qf_lib/data_providers/sp_global/sp_data_provider.py [hunk 8]
```diff
@@ -0,0 +1,359 @@
...
+    def price_field_to_str_map(self) -> Dict[PriceField, SPField]:
+        return {
+            PriceField.Open: SPField.OpenPrice,
+            PriceField.High: SPField.HighPrice,
+            PriceField.Low: SPField.LowPrice,
+            PriceField.Close: SPField.ClosePrice,
+            PriceField.Volume: SPField.Volume
+        }
+
+    def get_history(self, tickers: Union[SPTicker, Sequence[SPTicker]], fields: Union[SPField, Sequence[SPField]],
```
Comment (eirik-thorp) [inline qf_lib/data_providers/sp_global/sp_data_provider.py:89]: Will this ever return something else than a `QFDataArray`?

Comment (myrmarachne) [reply qf_lib/data_providers/sp_global/sp_data_provider.py:89]: Yes, whenever we have a single date, single ticker, single field the normalize_data_array will squeeze the array to am object with less dimensions

File: qf_lib/data_providers/sp_global/sp_dao.py [hunk 9]
```diff
@@ -44,12 +45,13 @@ def get_history(self, tickers: Union[SPTicker, Sequence[SPTicker]], fields: Unio
 
     @property
     @abstractmethod
-    def _supported_tables(self) -> list[SPField]:
+    def _supported_tables(self) -> list[str]:
         raise NotImplementedError
```
Comment (eirik-thorp) [inline qf_lib/data_providers/sp_global/sp_dao.py:49]: Add todo to maybe add SPTable Enum

**Pull request merged**